### ISSUE TITLE
act() now return always a Promise

### ISSUE THREAD

**Issue by GOI17** [testing-library/react-testing-library#1155]
# Description

Today I update my work project deps, and I noticed that the eslint throws me an error.

![image](https://user-images.githubusercontent.com/32525535/203448046-4fc44f3a-eab1-42ae-a37c-727c1eee4ffb.png)

At the beginning I thought that was my mistake, but I checked the code and I found that the error was dispatch because I'm calling `act(() => mycode())` as as sync code. 
I never had this problem before, so I checked my updated types, and I found this
` export const act: typeof reactAct extends undefined
     ? (callback: () => void) => void
     : typeof reactAct
 `, nothing change on testing-library types, but, react-dom updated their types,
 
 `export function act<T>(callback: () => T | Promise<T>): Promise<T>;`
 `export function act(callback: () => VoidOrUndefinedOnly): Promise<void>`
  
  on previous versions act had this types
  
`export function act(callback: () => Promise<void>): Promise<undefined>;`
`export function act(callback: () => VoidOrUndefinedOnly): void;`

Now my question, Will change the act type on testing-library or not?, act is still necessary if we already have the waitFor function?

## Versions
- react 18.2.0
- react-dom 18.2.0
- @types/react 18.0.23 -> 18.0.25
- @types/react-dom 18.0.7 -> 18.0.9
- @testing-library/react 13.14.0
- @testing-library/user-event 14.4.3

**Comment by icatalina** [comment]
PR introducing the change in `@types/react-dom`: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63186

**Comment by eps1lon** [comment]
Replaced by whatever issue is spawned from https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63186#issuecomment-1326546091.

Leaving open for visibility.

**Comment by DrewCerny** [comment]
Im wondering if this is related? I am now getting the dreaded not wrapped in act warning in hundreds of tests. I have read many times that render and fireEvent do not need to be wrapped in act calls. However, this is where the warnings are coming from. Is this because render and fireEvent are not awaiting their act calls? I have tested wrapping my render call in an await act() and it solves the warning problem...

**Comment by eps1lon** [comment]
@DrewCerny  We didn't change any runtime behavior recently.

**Comment by mrsants** [comment]
I was surprised to see that my typing does not inform a promise return in act, in the log of my tests I am seeing this:

`  ● Console

    console.error
      Warning: You called act(async () => ...) without await. This could lead to unexpected testing behaviour, interleaving multiple act calls and mixing their scopes. You should - await act(async () => ...);

      at printWarning (node_modules/react/cjs/react.development.js:209:30)
      at error (node_modules/react/cjs/react.development.js:183:7)
      at node_modules/react/cjs/react.development.js:2564:15
`

**Comment by eps1lon** [comment]
This was fixed in `@types/react@18.0.27`

**Comment by lll000111** [comment]
So... does this function always return a promise now?

The problem is if I "await" act() and the callback is synchronous I get errors from typescript-eslint-plugin:

```
ESLint: Unexpected `await` of a non-Promise (non-"Thenable") value.(@typescript-eslint/await-thenable)
```

- @types/react-test-renderer 18.0.0
- react-test-renderer 18.2.0
